File: typescript/sort/mergesort.ts




function mergeSort (arr : any[]) : any[]  {

    if (arr.length <= 1) {
        return arr
    }

    let middle = Math.floor(arr.length / 2)
    let leftArr = arr.slice(0,middle)
    let rightArr = arr.slice(middle)

    return mergeTwoSortedArray(mergeSort(leftArr), mergeSort(rightArr))

}




function mergeTwoSortedArray (left : any[], right : any[]) : any[] {

    let result : any[] = [];
    let leftIdx = 0;
    let rightIdx = 0;

    while (leftIdx < left.length && rightIdx < right.length) {


        if (left[leftIdx] < right[rightIdx]) {

            result.push( left[leftIdx] )
            leftIdx ++;
        } else {

            result.push( right[rightIdx] )
            rightIdx ++;
        }
    }

    return result.concat(left.slice(leftIdx).concat(right.slice(rightIdx)))

}


let sortedNumberArrayy: number[] =
    mergeSort([38, 27, 43, 10,5859, 5,5,5,5,5,4545,234,12,34,5,6,7,764325,3]);
console.log(sortedNumberArrayy);

File: python/sort/mergeSort.py



def merge_sort(arr) :
    
    if len(arr) <= 1:
        return arr
    
    mid = len(arr) // 2
    left = arr[:mid]
    right = arr[mid:]

    left = merge_sort(left)
    right = merge_sort(right)

    return merge_two_sorted_lists(left, right)

def merge_two_sorted_lists(leftArr,rightArr) : 

    sorted_list = []
    leftIdx = rightIdx = 0

    while leftIdx < len(leftArr) and rightIdx < len(rightArr) : 

        if leftArr[leftIdx] <= rightArr[rightIdx] : 
            sorted_list.append(leftArr[leftIdx])
            leftIdx += 1
        else :
            sorted_list.append(rightArr[rightIdx])
            rightIdx += 1

    while leftIdx < len(leftArr) : 
        sorted_list.append(leftArr[leftIdx])
        leftIdx += 1
    
    while rightIdx < len(rightArr) : 
        sorted_list.append(rightArr[rightIdx])
        rightIdx += 1
    
    return sorted_list


if __name__ == "__main__" :

    print(merge_sort([10,3,4,5,22,3,121,5]))